File: lab3/FileReverse.c
/*
 * FileReverse.c
 * Programmer: Tiancheng Fu
 * CMPS 12M
 * Date: July 18th, 2018
 * Cruzid: tfu6
 * The program will allows the user to enter a input and an output file. The program will read the words in the input file
 * and split the words from the whitespaces and empty lines in between them. Then the program will
 * print out the words contain in the input file which have been split in backward order into the user entered output file.
 * The FileReverse.c is built based on the FileReverse.java from Lab Assignment 2 and FileIO.c from the Lab Assignment 3 pdf file.
 * The FileReverse.c includes a void type function called stringReverse(), which will reverse the characters of String variable s.
 * The specific functions of the functions and codes are listed in the comments.
 *
 */
 //Insert the library header files
#include<stdio.h>
#include<stdlib.h>
#include<string.h>

  void stringReverse(char* s){//This void function will reverse the letters of a given word

  //Declare char array to store the letters from the given word s
  char* c [strlen(s)];

  //Insert the letters in given word s into char array c
  for(int i = 0; i < strlen(s); i++){
   c[i] = s[i];
  }
  int n = strlen(s);//Integer variable uses to store the length of the given word s

  //Reverse all the letters in the given word s by swap their position
  for(int j = n; j > strlen(s)/2; j--){

   s[strlen(s)-j] = c[j-1];//Copying the jth element from the left in c[] into the jth element from the right in word s
   s[j-1] = c[strlen(s)-j];//Copying the jth element from the right in c[] into the jth element from the left in word s

  }

}

//This function main will allow the user to enter an input file and an output file and split the words in the input file and use function stringReverse() to output those words in backward order in the user entered output file
int main(int argc, char* argv[]){
 FILE* in; /* File handle for input */
 FILE* out; /* File handle for output */

 char word[256]; /* Char array to store words from input file */
 /* Check command line for correct number of arguments */
 if( argc < 2 ){ /* Print out error statement if user entered less than two arguments */
 printf("Usage: %s <input file> <output file>\n", argv[0]);
 exit(EXIT_FAILURE);
 }

 /* Open input file for reading */
 in = fopen(argv[1], "r");
 if( in==NULL ){ /* Print out error statement if user entered input file is wrong */
 printf("Unable to read from file %s\n", argv[1]);
 exit(EXIT_FAILURE);
 }
 /* Open output file for writing */
 out = fopen(argv[2], "w");
 if( out==NULL ){ /* Print out error statement if user entered output file is wrong */
 printf("Unable to write to file %s\n", argv[2]);
 exit(EXIT_FAILURE);
 }

 /* Read words from input file and print on separate lines to output file after reversed them */
 while( fscanf(in, " %s", word) != EOF ){
  stringReverse(word);/* Reverse the word */
  fprintf(out, "%s\n", word); /* Print out the reversed word into the output file */
 }
 /* close input and output files and end the program */
 fclose(in);
 fclose(out);
 return(EXIT_SUCCESS);
}

File: lab2/FileReverse.java
/*
 * FileReverse.java
 * Programmer: Tiancheng Fu
 * CMPS 12M
 * Date: July 12th, 2018
 * Cruzid: tfu6
 * This program have one return type recursive method stringReverse and stringReverse2 which are based on reverseArray1,2,3 from pa1, and a main class that allow the user to enter a input file's name
 * and a output file's name. Then the program will read the words in the input and removes the blank lines and white spaces.
 * After that, the program will use the stringReverse and stringReverse2 methods to reverse the words and print them into the user named file.
 * The specific functions of the method and codes are listed in the rest of the comments. 
 * 
 */
//Importing packages to java
import java.io.*;
import java.util.*;
class FileReverse{
  
 public static void main(String[] args) throws IOException{//Throw IOExaception to the class so the program will keep running when there is a runtime error
   
   Scanner in = null;//Scanner to check inputs from the input file
   PrintWriter out = null;//PrintWriter use to output words to the output file
   String line = null;//String variable use to temporarily store a word from the input file
   String[] token = null;//String array use to store all the words in the input file
   
   int i, n;//Integer variables use to count the number of the words in the input file
   
 // Check number of command line arguments is at least 2
   if(args.length < 2){//If the number of command line arguments is less than 2 then stop the program
     System.out.println("Usage: FileCopy <input file> <output file>");
     System.exit(1);
 }
   
 // open files
   in = new Scanner(new File(args[0]));
   out = new PrintWriter(new FileWriter(args[1]));
   
 // read lines from in, extract and print tokens from each line
 while( in.hasNextLine() ){
 // trim leading and trailing spaces, then add one trailing space so
 // split works on blank lines
   line = in.nextLine().trim() + " ";
 // split line around white space
   token = line.split("\\s+");
 // print out tokens in reverse order
   n = token.length;
   
 for(i=0; i<n; i++){
   out.println(" "+stringReverse(token[i], token[i].length()));
  }
 }
 // close the output and input file
 in.close();
 out.close();
   
 }
 
  public static String stringReverse(String s, int n){//This recursive method will call a void method named stringReverse2 and return a String that is the reversal of the first n characters of String variable s
  
  //Delcare three char arrays to store the letters of the String s
  char [] c = new char [n];
  char [] d = new char [n];//Char array d stores the letters that have been reversed
  char [] e = new char [s.length()-n];//Char array e stores the letters that are not required to reverse
  
  //Insert the letters in String s that are required to reverse into char arrays d and c
  for(int i = 0; i < n; i++){
   d[i] = s.charAt(i);
   c[i] = s.charAt(i);
  }
  
  //Insert the letters in String s that are not required to reverse into char array e
  for(int j = n; j < s.length(); j++){
   e[j-n] = s.charAt(j); 
  }
  
   stringReverse2(c, n, d);//Call void method stringReverse2 to reverse the letters in the string s that are required to reverse
   s = String.valueOf(d) + String.valueOf(e);//Combine the letters in d and e into a word and insert it into String s
   
   return s;//return the word that has been reversed
  }
  
  public static void stringReverse2(char []c, int n, char []d){//This method will copy the leftmost n elements in c[] into the rightmost n positions in d[] in reverse order
    
    if(n>0){//If there still are letters left in the char array c that didn't get copy and swap its position then insert into char array d

    d[d.length-n] = c[n-1]; //Copying the nth element from the left in c[] into the nth element from the right in d[]
    
    stringReverse2(c, n-1, d); //Call the method itself again without copy and swap the elements that already got swapped
    
    }
 
}
 
}